### PR TITLE
Add ObjectSetPhase APIs

### PR DESCRIPTION
**Pull request by thetechnick**
Signed-off-by: Nico Schieder <nschieder@redhat.com>

### CODE REVIEW

File: apis/core/v1alpha1/clusterobjectset_types.go [hunk 1]
```diff
@@ -54,8 +55,8 @@ type ClusterObjectSetStatus struct {
 	// it will go away as soon as kubectl can print conditions!
 	// When evaluating object state in code, use .Conditions instead.
 	Phase ObjectSetStatusPhase `json:"phase,omitempty"`
-	// List of objects the controller has paused reconciliation on.
-	PausedFor []ObjectSetPausedObject `json:"pausedFor,omitempty"`
+	// Computed revision number to order revisions on a linear axis.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectset_types.go:58]:
```suggestion
	// Computed revision number, monotonically increasing. 
```

Comment (garrettthomaskth) [reply apis/core/v1alpha1/clusterobjectset_types.go:58]: Is that correct?

File: apis/core/v1alpha1/clusterobjectsetphase_types.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// ClusterObjectSetPhase is an internal API, allowing an ClusterObjectSet to delegate a single phase to another custom controller.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectsetphase_types.go:5]:
```suggestion
// ClusterObjectSetPhase is an internal API, allowing a ClusterObjectSet to delegate a single phase to another custom controller.
```

File: apis/core/v1alpha1/clusterobjectsetphase_types.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// ClusterObjectSets will create subordinate ClusterObjectSetPhases when `.class` within the phase specification is set.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectsetphase_types.go:6]:
```suggestion
// ClusterObjectSets will create subordinate ClusterObjectSetPhases when `.class` is set within the phase specification.
```

File: apis/core/v1alpha1/clusterobjectsetphase_types.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	// All probes need to succeed for a package to be considered Available.
+	// Failing probes will prevent the reconciliation of objects in later phases.
+	AvailabilityProbes []ObjectSetProbe `json:"availabilityProbes"`
+
+	ObjectSetTemplatePhase `json:",inline"`
+}
+
+// ClusterObjectSetPhaseStatus defines the observed state of a ClusterObjectSetPhase.
+type ClusterObjectSetPhaseStatus struct {
+	// Conditions is a list of status conditions ths object is in.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/clusterobjectsetphase_types.go:52]: I just want to confirm that it is not like the default `conditions` in kubernetes where it is a list of all status conditions the object is in and has been in

Comment (garrettthomaskth) [reply apis/core/v1alpha1/clusterobjectsetphase_types.go:52]: Judging from the example it isn't :)

File: apis/core/v1alpha1/commonobjectset_types.go [hunk 5]
```diff
@@ -35,6 +35,10 @@ type ObjectSetTemplateSpec struct {
 type ObjectSetTemplatePhase struct {
 	// Name of the reconcile phase. Must be unique within a ObjectSet.
 	Name string `json:"name"`
+	// If non empty, the ObjectSet controller will delegate phase reconciliation to another controller, by creating an ObjectSetPhase object.
+	// If set to the string "default" the build-in Package Operator ObjectSetPhase controller will reconcile the object in the same way the ObjectSet would.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/commonobjectset_types.go:39]:
```suggestion
	// If set to the string "default" the built-in Package Operator ObjectSetPhase controller will reconcile the object in the same way the ObjectSet would.
```

File: apis/core/v1alpha1/objectset_types.go [hunk 6]
```diff
@@ -55,8 +56,8 @@ type ObjectSetStatus struct {
 	// it will go away as soon as kubectl can print conditions!
 	// When evaluating object state in code, use .Conditions instead.
 	Phase ObjectSetStatusPhase `json:"phase,omitempty"`
-	// List of objects the controller has paused reconciliation on.
-	PausedFor []ObjectSetPausedObject `json:"pausedFor,omitempty"`
+	// Computed revision number to order revisions on a linear axis.
```
Comment (garrettthomaskth) [inline apis/core/v1alpha1/objectset_types.go:59]:
```suggestion
	// Computed revision number, monotonically increasing.
```

Comment (garrettthomaskth) [reply apis/core/v1alpha1/objectset_types.go:59]: :D

File: apis/core/v1alpha1/clusterobjectsetphase_types.go (hunk 4 above)
Comment (thetechnick) [reply apis/core/v1alpha1/clusterobjectsetphase_types.go:52]: @garrettthomaskth Checkout https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
Conditions as a status history has been discontinued for a long time now.

**Pull request merged**